wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/environment/server/CheEnvironmentEngine.java

```diff
@@ -0,0 +1,954 @@
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import static java.lang.String.format;
+import static org.eclipse.che.api.machine.server.event.InstanceStateEvent.Type.DIE;
+import static org.eclipse.che.api.machine.server.event.InstanceStateEvent.Type.OOM;
+import static org.eclipse.che.dto.server.DtoFactory.newDto;
+import static org.slf4j.LoggerFactory.getLogger;
+
+/**
+ * Facade for implementation specific operations with environment runtimes.
+ *
+ * @author Alexander Garagatyi
+ * @author Yevhenii Voevodin
+ */
+@Singleton
+public class CheEnvironmentEngine {
+
+    private static final Logger LOG = getLogger(CheEnvironmentEngine.class);
+
+    private final Map<String, EnvironmentHolder> environments;
+    private final StripedLocks                   stripedLocks;
+    private final SnapshotDao                    snapshotDao;
+    private final File                           machineLogsDir;
+    private final MachineInstanceProviders       machineInstanceProviders;
+    private final int                            defaultMachineMemorySizeMB;
+    private final EventService                   eventService;
+
+    private volatile boolean isPreDestroyInvoked;
+
+    @Inject
+    public CheEnvironmentEngine(SnapshotDao snapshotDao,
+                                MachineInstanceProviders machineInstanceProviders,
+                                @Named("machine.logs.location") String machineLogsDir,
+                                @Named("machine.default_mem_size_mb") int defaultMachineMemorySizeMB,
+                                EventService eventService) {
+        this.eventService = eventService;
+        this.environments = new ConcurrentHashMap<>();
+        this.snapshotDao = snapshotDao;
+        this.machineInstanceProviders = machineInstanceProviders;
+        this.machineLogsDir = new File(machineLogsDir);
+        this.defaultMachineMemorySizeMB = defaultMachineMemorySizeMB;
+        // 16 - experimental value for stripes count, it comes from default hash map size
+        this.stripedLocks = new StripedLocks(16);
+        eventService.subscribe(new MachineCleaner());
+    }
+
+    /**
+     * Returns all machines from environment of specific workspace.
+     *
+     * @param workspaceId
+     *         ID of workspace that owns environment machines
+     * @return list of machines
+     * @throws NotFoundException
+     *         if environment is not running
+     */
+    public List<Instance> getMachines(String workspaceId) throws NotFoundException {
+        EnvironmentHolder environment;
+        try (StripedLocks.ReadLock lock = stripedLocks.acquireReadLock(workspaceId)) {
+            environment = environments.get(workspaceId);
+            if (environment == null) {
+                throw new NotFoundException("Environment with ID " + workspaceId + " is not found");
+            }
+        }
+        return new ArrayList<>(environment.machines);
```

Comment:
While you copy machines list, machines themself are the same instances, it looks like you should copy machines as well + i'd return just from try, as it won't produce any stale data issues.